sklearn/datasets/value_dropper.py

```diff
@@ -0,0 +1,404 @@
+                                     " target labels as keys, the values of "
+                                     "the dict should either be a single float"
+                                     " or an array of shape (n_features,). "
+                                     "%r was passed" % (val))
+
+            if drop_probs.sum() >= 1:
+                raise ValueError("The sum of all probabilities in the "
+                                 "missing_distribution should sum up to less "
+                                 "than 1. "
+                                 "The sum was found to be %0.8f"
+                                 % drop_probs.sum())
+
+        else:
+            # For MCAR
+
+            # Validate and convert the missing_distribution dict into a
+            # 1D probability distribution along the features
+
+            drop_probs = np.zeros(n_features,
+                                  dtype=np.float64).reshape((1, n_features))
+
+            if isinstance(self.missing_distribution, (float, np.floating)):
+                drop_probs[:] = self.missing_distribution / n_features
+            elif isinstance(self.missing_distribution,
+                            (list, tuple, np.ndarray)):
+                # TODO validate shape / presence of NaN/None etc.
+                drop_probs[:] = self.missing_distribution
+            else:
+                raise ValueError("missing_distribution must be a float or "
+                                 " 1D vector (list, tuple or np.ndarray) of "
+                                 "shape (n_features,) or dict of 1D vector / "
+                                 "floats. %r was passed"
+                                 % self.missing_distribution)
+
+        missing_rate = drop_probs.sum()
+        current_n_missing = np.count_nonzero(missing_mask)
+        required_n_missing = int(missing_rate * n_elements)
+        n_more_missing = required_n_missing - current_n_missing
+
+        if current_n_missing > required_n_missing:
+            raise ValueError("There are %d existing missing values, "
+                             "which is > %d that is "
+                             "expected to be missing as given by the "
+                             "missing_distribution %s."
+                             % (current_n_missing, required_n_missing,
+                                missing_distribution))
+
+        drop_counts = (drop_probs * n_elements).astype(int)
+
+        def generate_sample_masks():
+            if missing_type == 'mcar':
+                # For MCAR, do not consider target values, select all samples
+                # in a single set
+                yield np.ones(n_samples, dtype=bool)
+            else:
+                # For classification, NMAR, consider missing distribution
+                # based on class labels as subsets within data
+                for label in labels:
+                    yield y == label
```

Comment:
I think you can split your transform into two private functions for that part as we discussed.